distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Split_Tokenize.enso

```diff
@@ -0,0 +1,246 @@
+tokenize_to_columns table input_column_name pattern case_sensitivity column_count on_problems =
+    expect_text_column table input_column_name <|
+        fan_out_to_columns table input_column_name (ignore_nothing (_.tokenize pattern case_sensitivity)) column_count on_problems
+
+## PRIVATE
+   Tokenizes a column of text into a set of new rows using a regular
+   expression.
+   See `Table.tokenize_to_rows`.
+tokenize_to_rows : Table -> Text | Integer -> Text -> Case_Sensitivity -> Problem_Behavior -> Table
+tokenize_to_rows table input_column_name pattern="." case_sensitivity=Case_Sensitivity.Sensitive on_problems=Report_Error =
+    _ = [on_problems]
+    expect_text_column table input_column_name <|
+        fan_out_to_rows table input_column_name (ignore_nothing (_.tokenize pattern case_sensitivity))
+
+## PRIVATE
+   Transform a table by transforming a column into a set of columns. Takes a
+   function that maps a single element of the input column to a vector of output
+   values. The original column is replaced by the new columns.
+
+   Arguments:
+   - table: The table to transform.
+   - input_column: The column to transform.
+   - function: A function that transforms a single element of `input_column`
+     to multiple values.
+fan_out_to_columns : Table -> Text | Integer -> (Any -> Vector Any) -> Auto | Integer -> Problem_Behavior -> Table | Nothing
+fan_out_to_columns table input_column_name function column_count=Auto on_problems=Report_Error =
+    input_column = table.get input_column_name
+    problem_builder = Problem_Builder.new
+    new_columns_unrenamed = map_columns_to_multiple input_column function column_count problem_builder
+    new_columns = rename_new_columns table new_columns_unrenamed problem_builder
+    new_table = replace_column_with_columns table input_column new_columns
+    problem_builder.attach_problems_after on_problems new_table
+
+## PRIVATE
+   Transform a column by applying the given function to the values in the
+   column. The function produces multiple outputs, so each row is duplicated,
+   with each row getting a distinct output value in place of the original
+   input value. The other column values are just duplicated.
+
+   Arguments:
+   - table: The table to transform.
+   - input_column: The column to transform.
+   - function: A function that transforms a single element of `input_column`
+     to multiple values.
+fan_out_to_rows : Table -> Text | Integer -> (Any -> Vector Any) -> Table
+fan_out_to_rows table input_column_name function =
+    input_column = table.at input_column_name
+    input_storage = input_column.java_column.getStorage
+    num_input_rows = input_storage.size
+
+    # Accumulates the output of the output column values.
+    output_column_builder = make_string_builder 0
+    # Accumulates repeated position indices for the order mask.
+    order_mask_positions = Vector.new_builder 0
```

Comment:
```suggestion
    # The functions we use return at least one element for each input element, so the result will have at least `input_column.length` entries - so we can initialize the builders with that to avoid early reallocations.
    initial_size = input_column.length
    # Accumulates the output of the output column values.
    output_column_builder = make_string_builder initial_size
    # Accumulates repeated position indices for the order mask.
    order_mask_positions = Vector.new_builder initial_size
```

Reply:
I did it this way because I want this code to work unchanged for other functions that might return 0 elements. (Tokenize can return 0 elements anyway, if there are no matches.)

Reply:
I see. Well, still `initial_size` of 0 seems weird to me, I'd pick at least 8/16, so that it can start growing a bit faster.

Also keep in mind that the `initial_size` of the builder can be larger than the effective size - the only downside is that we may allocate too much because of it - but if we are doing less reallocations I guess it may sometimes incur less cost all in all.

What about making `initial_size` a parameter of this function, defaulting to `column.length` but possible to override it if we need to use it in other function which prefer a smaller starting size?

Reply:
As for tokenize - interesting! I did not think of it. I guess this is more of a design question (cc: @jdunkerley) but I'm wondering if in such a case that there are no matches, shouldn't `tokenize` return a single row with `Nothing` in it? One can always then filter out these rows with `Filter_Condition.Not_Nothing`, but I imagine we may not always want to lose these rows.

Reply:
Ah, I didn't realize the initial size was different from the effective size -- I'll start with initial_size for sure.

I think it makes sense for `tokenize` to return `[]` no for no matches; unlike `split`, it's return value consists of the matches, possibly grouped, and if there are no matches, it should return an empty vector.

Reply:
`tokenize` itself - definitely!

I meant `tokenize_to_rows`. There we are losing rows whose `tokenize = []`. And I'm not sure if instead we should have a single row with `Nothing`. A bit like with a `LEFT JOIN` - I get the rows that matched as many times as there are matches, but if there was 0 matches, I still get one row with NULLs.

Reply:
I see your point; I think adding Nothing could be an option, but we should default to not adding it -- just like LEFT is an explicitly option for JOIN.

Reply:
Yeah, I think ideally we should have a parameter like `keep_unmatched`. I agree it may default to `False`.

I think what we have is fine for now, but I'd consider if it's worth adding a follow up ticket, with lower priority perhaps.

@jdunkerley shall we add a ticket? Or just ignore for now and see how it's used? That's also an option I guess.